### DISCUSSION TITLE
Custom response on source connector HTTP Listener

### DISCUSSION THREAD

**Discussion by cw85865** [nextgenhealthcare/connect#6350] [Q&A] [answered]
I have a channel that is intended to receive FHIR JSON messages with the source connector setup as a HTTP Listener. Because we queue everything, what I would like to do is configure things so that Mirth can respond to the sender before the message is processed ("Source Queue" set to "On"), but with a custom response instead of an auto-generate response. Is this possible to do?

**Comment by kirbykn2** [comment]
From: https://chatgpt.com/g/g-XG1Dzv32o-interop-guru

UPDATED RESPONSE:

  Potential solution for Immediate Custom Response on Queued Messages
To achieve an immediate custom response for queued messages, we need to use a different approach with the HTTP Listener itself:

Use return in the Source Transformer:

The HTTP Listener allows you to specify a custom response by returning a value directly from the Source Transformer.
By adding logic to the Source Transformer, you can return a custom response to the sender as soon as the message is received and queued, without waiting for the full processing in the channel.
Custom Response Script in the Source Transformer:

You can set up the Source Transformer with a short script to define a custom response that will be sent immediately.

This script can be something like:

javascript
Copy code
// Custom response message sent immediately after message is queued
var response = {
    status: "queued",
    message: "Your message has been received and queued for processing.",
    transactionID: UUIDGenerator.getUUID() // Optionally add a transaction ID
};

// Return JSON as response
return JSON.stringify(response);
This code defines a JSON response that will be immediately sent back to the sender, acknowledging that the message has been queued for processing.

Response Status Code Configuration:

To customize the HTTP status code (such as sending an HTTP 202 Accepted status), configure this under the Response Settings in the HTTP Listener properties.
Under Response > Response Status in the HTTP Listener settings, set the HTTP Status Code to 202 or any code that matches your requirements.
Summary
By using a return statement in the Source Transformer, you’ll achieve an immediate custom response back to the HTTP sender as soon as the message is queued, without waiting for the complete processing. This approach works well for scenarios where you want quick acknowledgment, especially in queued or asynchronous processing contexts in Mirth Connect.

On Tue, Nov 12, 2024 at 12:39 PM Christopher Williams <
***@***.***> wrote:

> I have a channel that is intended to receive FHIR JSON messages with the
> source connector setup as a HTTP Listener. Because we queue everything,
> what I would like to do is configure things so that Mirth can respond to
> the sender before the message is processed ("Source Queue" set to "On"),
> but with a custom response instead of an auto-generate response. Is this
> possible to do?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/nextgenhealthcare/connect/discussions/6350>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/APRXWD2QLEEL53M3LCX6BXT2AI4NNAVCNFSM6AAAAABRUTLKKOVHI2DSMVQWIX3LMV43ERDJONRXK43TNFXW4OZXGQ3TCNJVGU>
> .
> You are receiving this because you are subscribed to this thread.Message
> ID: ***@***.***>
>


-- 
Best,

Kirby Knight
| 231.735.4650 | ***@***.***

**Reply by cw85865** [reply]
I have tried many variations of this and have gotten nowhere. If `Source Queue` is set to `On`, then no JS code anywhere has an impact on the response (with `Response` set to `Auto-generate (Before processing)`). If `Source Queue` is set to `Off`, the only thing that actually changes the response is by using the postprocessor script and returning the value we need. None of the changes I've made to the source filter or the source transformer have any impact on the response, regardless of the setting of `Source Queue` and with trying all of the possible settings for `Response`.

**Comment by kirbykn2** [comment]
You may need to use a two channel design.

On Tue, Nov 12, 2024 at 1:26 PM Christopher Williams <
***@***.***> wrote:

> I have tried many variations of this and have gotten nowhere. If Source
> Queue is set to On, then no JS code anywhere has an impact on the
> response (with Response set to Auto-generate (Before processing)). If Source
> Queue is set to Off, the only thing that actually changes the response is
> by using the postprocessor script and returning the value we need. None of
> the changes I've made to the source filter or the source transformer have
> any impact on the response, regardless of the setting of Source Queue and
> with trying all of the possible settings for Response.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/nextgenhealthcare/connect/discussions/6350#discussioncomment-11230382>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/APRXWDZHFTRDKAT4QMC36DL2AJB4NAVCNFSM6AAAAABRUTLKKOVHI2DSMVQWIX3LMV43URDJONRXK43TNFXW4Q3PNVWWK3TUHMYTCMRTGAZTQMQ>
> .
> You are receiving this because you commented.Message ID:
> ***@***.***
> com>
>


-- 
Best,

Kirby Knight
| 231.735.4650 | ***@***.***

**Comment by kirbykn2** [comment]
- Immediate custom response to the sender, controlled by *Channel 1*.
- Full processing of the message with queuing in *Channel 2*.

On Tue, Nov 12, 2024 at 2:43 PM Kirby Knight ***@***.***> wrote:

> You may need to use a two channel design.
>
> On Tue, Nov 12, 2024 at 1:26 PM Christopher Williams <
> ***@***.***> wrote:
>
>> I have tried many variations of this and have gotten nowhere. If Source
>> Queue is set to On, then no JS code anywhere has an impact on the
>> response (with Response set to Auto-generate (Before processing)). If Source
>> Queue is set to Off, the only thing that actually changes the response
>> is by using the postprocessor script and returning the value we need. None
>> of the changes I've made to the source filter or the source transformer
>> have any impact on the response, regardless of the setting of Source
>> Queue and with trying all of the possible settings for Response.
>>
>> —
>> Reply to this email directly, view it on GitHub
>> <https://github.com/nextgenhealthcare/connect/discussions/6350#discussioncomment-11230382>,
>> or unsubscribe
>> <https://github.com/notifications/unsubscribe-auth/APRXWDZHFTRDKAT4QMC36DL2AJB4NAVCNFSM6AAAAABRUTLKKOVHI2DSMVQWIX3LMV43URDJONRXK43TNFXW4Q3PNVWWK3TUHMYTCMRTGAZTQMQ>
>> .
>> You are receiving this because you commented.Message ID:
>> ***@***.***
>> .com>
>>
>
>
> --
> Best,
>
> Kirby Knight
> | 231.735.4650 | ***@***.***
>


-- 
Best,

Kirby Knight
| 231.735.4650 | ***@***.***

**Reply by pacmano1** [reply, marked as answer]
Please see https://forums.mirthproject.io/forum/mirth-connect/support/18826-http-listener-custom-response-with-source-queue-on?p=104294#post104294

Reminder to check there also.

**Reply by pacmano1** [reply]
@kirbykn2 , I would strike your chatGPT answer entirely.
